internal/controller/dispatcher.go

```diff
@@ -0,0 +1,258 @@
+			}
+		}
+		// set queued/idle status only after adding finalizers
+		return r.updateStatus(ctx, appWrapper, mcadv1beta1.Queued, mcadv1beta1.Idle)
+
+	case mcadv1beta1.Queued:
+		// Propagate most recent queuing decision to AppWrapper's Queued Condition
+		if decision, ok := r.Decisions[appWrapper.UID]; ok {
+			meta.SetStatusCondition(&appWrapper.Status.Conditions, metav1.Condition{
+				Type:    string(mcadv1beta1.Queued),
+				Status:  metav1.ConditionTrue,
+				Reason:  string(decision.reason),
+				Message: decision.message,
+			})
+			if r.Status().Update(ctx, appWrapper) == nil {
+				// If successfully propagated, remove from in memory map
+				delete(r.Decisions, appWrapper.UID)
+			}
+		}
+
+		if meta.FindStatusCondition(appWrapper.Status.Conditions, string(mcadv1beta1.Queued)) == nil {
+			// Absence of Queued Condition strongly suggests AppWrapper is new; trigger dispatch and a short requeue
+			r.triggerDispatch()
+			return ctrl.Result{RequeueAfter: deletionDelay}, nil
+		} else {
+			return ctrl.Result{RequeueAfter: queuedDelay}, nil
+		}
+
+	case mcadv1beta1.Running:
+		switch appWrapper.Status.Step {
+		case mcadv1beta1.Dispatching:
+			if r.MultiClusterMode {
+				// TODO: Multicluster. This is where we create the placement that maps
+				// the appWrapper to the execution cluster. Only if the placement creation is successful
+				// will we do the update of the status to running/creating.
+				return ctrl.Result{}, fmt.Errorf("unimplemented multi-cluster synch path")
+			}
+			return r.updateStatus(ctx, appWrapper, mcadv1beta1.Running, mcadv1beta1.Creating)
+
+		case mcadv1beta1.Deleted:
+			if r.MultiClusterMode {
+				// TODO: Multicluster. This is where we delete the placement that maps
+				// the appWrapper to the execution cluster. Only after the placement is
+				// completely deleted will we proceed to resetting the status to queued/idle
+				return ctrl.Result{}, fmt.Errorf("unimplemented multi-cluster synch path")
+			}
+
+			// reset status to queued/idle
+			appWrapper.Status.Restarts += 1
+			appWrapper.Status.RequeueTimestamp = metav1.Now() // overwrite requeue decision time with completion time
+			msg := "Requeued by MCAD"
+			if decision, ok := r.Decisions[appWrapper.UID]; ok && decision.reason == mcadv1beta1.QueuedRequeue {
+				msg = fmt.Sprintf("Requeued because %s", decision.message)
+			}
+			meta.SetStatusCondition(&appWrapper.Status.Conditions, metav1.Condition{
+				Type:    string(mcadv1beta1.Queued),
+				Status:  metav1.ConditionTrue,
+				Reason:  string(mcadv1beta1.QueuedRequeue),
+				Message: msg,
+			})
+			res, err := r.updateStatus(ctx, appWrapper, mcadv1beta1.Queued, mcadv1beta1.Idle)
+			if err == nil {
+				delete(r.Decisions, appWrapper.UID)
+				r.triggerDispatch()
+			}
+			return res, err
+		}
+
+	case mcadv1beta1.Succeeded:
+		switch appWrapper.Status.Step {
+		case mcadv1beta1.Idle:
+			r.triggerDispatch()
+			return ctrl.Result{}, nil
+		}
+
+	case mcadv1beta1.Failed:
```

Comment:
We should transition `Failed, deleted` to `Failed, idle` by removing the placement object in multi-cluster mode.

Reply:
done